File: js/test/varint.spec.ts
import { suite, test } from "mocha-typescript";
import { expect } from "chai";
import { Varint } from "../src/varint";
import { VarintExample } from "./varint_examples";

@suite class VarintEncode {
  static examples: VarintExample[];

  static before() {
    return VarintExample.loadAll(examples => this.examples = examples);
  }

  @test "encodes valid examples"() {
    for (let example of VarintEncode.examples) {
      if (!example.success) {
        continue;
      }

      expect(Varint.encode(example.value)).to.eql(example.encoded);
    }

    // 2**53-1 MAX (presently capped by JS integer precision)
    expect(Varint.encode(Varint.MAX)).to.eql(new Uint8Array([0x80, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0x1F]));
  }

  @test "throws TypeError for non-integer param"() {
    expect(() => Varint.encode(0.5)).to.throw(TypeError);
  }

  @test "throws RangeError for negative param"() {
    expect(() => Varint.encode(-1)).to.throw(RangeError);
  }

  @test "throws RangeError for param larger than Varint.MAX"() {
    expect(() => Varint.encode(Varint.MAX + 1)).to.throw(RangeError);
  }
}

@suite class VarintDecode {
  static examples: VarintExample[];

  static before() {
    return VarintExample.loadAll(examples => this.examples = examples);
  }

  @test "decodes valid examples"() {
    for (let example of VarintEncode.examples) {
      if (example.success) {
        expect(Varint.decode(example.encoded)[0]).to.eql(example.value);
      } else {
        expect(() => Varint.decode(example.encoded)).to.throw(Error);
      }
    }

    // Serialization of Varint.MAX
    expect(Varint.decode(new Uint8Array([0x80, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0x1F]))[0]).to.eql(Varint.MAX);
  }

  @test "throws Error for empty Uint8Array"() {
    expect(() => Varint.decode(new Uint8Array(0))).to.throw(Error);
  }

  @test "throws RangeError for params larger than Varint.MAX"() {
    // Serialization of Varint.MAX + 1
    expect(() => Varint.decode(new Uint8Array([0x80, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x32]))).to.throw(RangeError);
  }
}

File: ruby/spec/zser/varint_spec.rb
# encoding: binary
# frozen_string_literal: true

RSpec.describe Zser::Varint do
  describe ".encode" do
    context "varint.tjson examples" do
      it "encodes examples successfully" do
        VarintExample.load_file.each do |ex|
          next unless ex.success
          expect(described_class.encode(ex.value)).to eq ex.encoded
        end
      end
    end

    it "raises TypeError if given a non-Integer" do
      expect { described_class.encode(0.5) }.to raise_error TypeError
    end

    it "raises ArgumentError if given a negative value" do
      expect { described_class.encode(-1) }.to raise_error ArgumentError
    end

    it "raises ArgumentError if given Integer is larger than 64-bit unsigned" do
      expect { described_class.encode(Zser::Varint::MAX + 1) }.to raise_error ArgumentError
    end
  end

  describe ".decode" do
    it "decodes examples successfully" do
      VarintExample.load_file.each do |ex|
        if ex.success
          expect(described_class.decode(ex.encoded)).to eq [ex.value, ""]
        else
          expect { described_class.decode(ex.encoded) }.to raise_error Zser::ParseError
        end
      end
    end

    it "raises ArgumentError on an empty string" do
      expect { described_class.decode("") }.to raise_error ArgumentError
    end

    it "raises TypeError if given a non-string type" do
      expect { described_class.decode(42) }.to raise_error TypeError
    end

    it "raises Zser::TruncatedMessageError if input is truncated" do
      expect { described_class.decode("\x02") }.to raise_error Zser::TruncatedMessageError
      expect { described_class.decode("\x00\xFF") }.to raise_error Zser::TruncatedMessageError
    end
  end
end
